cow_instrument/ComponentInfo.h

```diff
@@ -68,75 +90,87 @@ ComponentInfo<InstTree>::ComponentInfo(
     : m_detectorInfo(std::move(detectorInfo)),
       m_instrumentTree(m_detectorInfo->const_instrumentTree()),
       m_componentToDetectorIndex(m_instrumentTree.componentSize(), -1),
-      m_componentToPathIndex(m_instrumentTree.componentSize(), -1) {
+      m_componentToPathIndex(m_instrumentTree.componentSize(), -1),
+      m_componentToBranchNodeIndex(m_instrumentTree.componentSize(), -1),
+      m_positions(std::make_shared<std::vector<Eigen::Vector3d>>(
+          m_instrumentTree.nBranchNodeComponents())),
+      m_rotations(std::make_shared<std::vector<Eigen::Quaterniond>>(
+          m_instrumentTree.nBranchNodeComponents())),
+      m_branchNodeComponentIndexes(std::make_shared<std::vector<size_t>>(
+          m_instrumentTree.branchNodeComponentIndexes())) {
+
+  init();
+}
+
+template <typename InstrTree> void ComponentInfo<InstrTree>::init() {
+  // TODO. Do this without copying everything!
+  std::vector<Eigen::Vector3d> allComponentPositions =
+      m_instrumentTree.startPositions();
+  std::vector<Eigen::Quaterniond> allComponentRotations =
+      m_instrumentTree.startRotations();
+
+  size_t i = 0;
+  for (auto &compIndex : (*m_branchNodeComponentIndexes)) {
+    (*m_positions)[i] = allComponentPositions[compIndex];
+    (*m_rotations)[i] = allComponentRotations[compIndex];
+    ++i;
+  }
   makeInvertedMaps();
 }
 
 template <typename InstTree> void ComponentInfo<InstTree>::makeInvertedMaps() {
 
-  // Invert the map.
+  // Invert the map. Allows us to use component index as a key.
   auto detToComponent = m_instrumentTree.detectorComponentIndexes();
   auto pathToComponent = m_instrumentTree.pathComponentIndexes();
+  auto branchNodeToComponent = m_instrumentTree.branchNodeComponentIndexes();
   for (size_t i = 0; i < detToComponent.size(); ++i) {
     m_componentToDetectorIndex[detToComponent[i]] = i;
   }
   for (size_t i = 0; i < pathToComponent.size(); ++i) {
     m_componentToPathIndex[pathToComponent[i]] = i;
   }
+  for (size_t i = 0; i < branchNodeToComponent.size(); ++i) {
+    m_componentToBranchNodeIndex[branchNodeToComponent[i]] = i;
+  }
 }
 
-
 template <typename InstTree>
 Eigen::Vector3d ComponentInfo<InstTree>::position(size_t componentIndex) const {
 
-  const std::vector<size_t> componentIndexes =
-      m_instrumentTree.nextLevelIndexes(componentIndex);
-  const auto nextLevelSize = componentIndexes.size();
-  if (nextLevelSize > 0) {
-    // We are dealing with a composite
-    auto pos = Eigen::Vector3d{0, 0, 0};
-    for (auto &index : componentIndexes) {
-      pos += this->position(index);
-    }
-    pos /= nextLevelSize;
-    return pos;
+  auto detIndex = m_componentToDetectorIndex[componentIndex];
```

Comment:
That would be handy. I think that the sort would need to be applied at the `FlatTree` level because it'd be important to ensure that all arrays are re ordered as the ComponentProxy vector is sorted.